File: packages/shared-contracts/src/enums.ts
/**
 * 跨服务枚举 — 单一真相源镜像自 spec/enums.json。
 * consistency 测试校验本文件与 spec、Python 绑定三者一致。
 */

export const PROVIDERS = ['tushare', 'akshare', 'realtime_sina', 'realtime_tencent'] as const;
export type Provider = (typeof PROVIDERS)[number];

export const PROVIDER_STATUSES = ['ok', 'error', 'unknown'] as const;
export type ProviderStatus = (typeof PROVIDER_STATUSES)[number];

export const DATASETS = [
  'price',
  'adj_factor',
  'daily_basic',
  'fundamental',
  'fina_indicator',
  'northbound',
  'index_ohlcv',
  'index_weight',
  'sw_industry',
] as const;
export type Dataset = (typeof DATASETS)[number];

export const JOB_TYPES = [
  'cache_build',
  'incremental',
  'train',
  'signal_gen',
  'paper_run',
  'news_fetch',
] as const;
export type JobType = (typeof JOB_TYPES)[number];

export const JOB_STATUSES = ['queued', 'running', 'paused', 'done', 'failed', 'canceled'] as const;
export type JobStatus = (typeof JOB_STATUSES)[number];

export const JOB_TRIGGERS = ['manual', 'schedule', 'onboarding'] as const;
export type JobTrigger = (typeof JOB_TRIGGERS)[number];

export const BOARDS = ['sh', 'sz', 'bj'] as const;
export type Board = (typeof BOARDS)[number];

export const PORTFOLIOS = ['model', 'personal'] as const;
export type Portfolio = (typeof PORTFOLIOS)[number];

export const SIGNAL_ACTIONS = ['buy', 'sell', 'hold'] as const;
export type SignalAction = (typeof SIGNAL_ACTIONS)[number];

export const TRADE_SIDES = ['buy', 'sell'] as const;
export type TradeSide = (typeof TRADE_SIDES)[number];

export const TRADE_REASONS = [
  'signal',
  'stop_loss',
  'take_profit',
  'rank_out',
  'market_filter',
  'manual',
] as const;
export type TradeReason = (typeof TRADE_REASONS)[number];

export const LOG_LEVELS = ['debug', 'info', 'warn', 'error'] as const;
export type LogLevel = (typeof LOG_LEVELS)[number];

export const ONBOARDING_STEPS = [
  'welcome',
  'select_source',
  'credential',
  'test',
  'build_cache',
  'done',
] as const;
export type OnboardingStep = (typeof ONBOARDING_STEPS)[number];

export const MODEL_TYPES = ['elasticnet'] as const;
export type ModelType = (typeof MODEL_TYPES)[number];

export const MODEL_STATUSES = ['draft', 'running', 'archived'] as const;
export type ModelStatus = (typeof MODEL_STATUSES)[number];

// 回测打分口径:auto=镜像实盘(有激活模型用模型,否则启用的自定义因子,再否则等权);
// equal_weight/model/custom 为强制档。落库的 scoring 为解析后的实际口径(engine 永不回传 auto)。
export const BACKTEST_SCORINGS = ['auto', 'equal_weight', 'model', 'custom'] as const;
export type BacktestScoring = (typeof BACKTEST_SCORINGS)[number];

/** 镜像所有枚举,供 consistency 测试逐项与 spec/enums.json 比对。 */
export const ENUM_MIRROR: Record<string, readonly string[]> = {
  Provider: PROVIDERS,
  ProviderStatus: PROVIDER_STATUSES,
  Dataset: DATASETS,
  JobType: JOB_TYPES,
  JobStatus: JOB_STATUSES,
  JobTrigger: JOB_TRIGGERS,
  Board: BOARDS,
  Portfolio: PORTFOLIOS,
  SignalAction: SIGNAL_ACTIONS,
  TradeSide: TRADE_SIDES,
  TradeReason: TRADE_REASONS,
  LogLevel: LOG_LEVELS,
  OnboardingStep: ONBOARDING_STEPS,
  ModelType: MODEL_TYPES,
  ModelStatus: MODEL_STATUSES,
  BacktestScoring: BACKTEST_SCORINGS,
};

File: packages/shared-contracts/python/sinan_contracts/enums.py
"""跨服务枚举 Python 绑定 —— 镜像自 spec/enums.json。"""

from __future__ import annotations

from enum import Enum


class _StrEnum(str, Enum):
    def __str__(self) -> str:  # 让序列化/日志输出取值而非 "Class.MEMBER"
        return str(self.value)


class Provider(_StrEnum):
    TUSHARE = "tushare"
    AKSHARE = "akshare"
    REALTIME_SINA = "realtime_sina"
    REALTIME_TENCENT = "realtime_tencent"


class ProviderStatus(_StrEnum):
    OK = "ok"
    ERROR = "error"
    UNKNOWN = "unknown"


class Dataset(_StrEnum):
    PRICE = "price"
    ADJ_FACTOR = "adj_factor"
    DAILY_BASIC = "daily_basic"
    FUNDAMENTAL = "fundamental"
    FINA_INDICATOR = "fina_indicator"
    NORTHBOUND = "northbound"
    INDEX_OHLCV = "index_ohlcv"
    INDEX_WEIGHT = "index_weight"
    SW_INDUSTRY = "sw_industry"


class JobType(_StrEnum):
    CACHE_BUILD = "cache_build"
    INCREMENTAL = "incremental"
    TRAIN = "train"
    SIGNAL_GEN = "signal_gen"
    PAPER_RUN = "paper_run"
    NEWS_FETCH = "news_fetch"


class JobStatus(_StrEnum):
    QUEUED = "queued"
    RUNNING = "running"
    PAUSED = "paused"
    DONE = "done"
    FAILED = "failed"
    CANCELED = "canceled"


class JobTrigger(_StrEnum):
    MANUAL = "manual"
    SCHEDULE = "schedule"
    ONBOARDING = "onboarding"


class Board(_StrEnum):
    SH = "sh"
    SZ = "sz"
    BJ = "bj"


class Portfolio(_StrEnum):
    MODEL = "model"
    PERSONAL = "personal"


class SignalAction(_StrEnum):
    BUY = "buy"
    SELL = "sell"
    HOLD = "hold"


class TradeSide(_StrEnum):
    BUY = "buy"
    SELL = "sell"


class TradeReason(_StrEnum):
    SIGNAL = "signal"
    STOP_LOSS = "stop_loss"
    TAKE_PROFIT = "take_profit"
    RANK_OUT = "rank_out"
    MARKET_FILTER = "market_filter"
    MANUAL = "manual"


class LogLevel(_StrEnum):
    DEBUG = "debug"
    INFO = "info"
    WARN = "warn"
    ERROR = "error"


class OnboardingStep(_StrEnum):
    WELCOME = "welcome"
    SELECT_SOURCE = "select_source"
    CREDENTIAL = "credential"
    TEST = "test"
    BUILD_CACHE = "build_cache"
    DONE = "done"


class ModelType(_StrEnum):
    ELASTICNET = "elasticnet"


class ModelStatus(_StrEnum):
    DRAFT = "draft"
    RUNNING = "running"
    ARCHIVED = "archived"


# 供 consistency 测试逐项比对 spec/enums.json。
ENUM_MIRROR: dict[str, list[str]] = {
    "Provider": [e.value for e in Provider],
    "ProviderStatus": [e.value for e in ProviderStatus],
    "Dataset": [e.value for e in Dataset],
    "JobType": [e.value for e in JobType],
    "JobStatus": [e.value for e in JobStatus],
    "JobTrigger": [e.value for e in JobTrigger],
    "Board": [e.value for e in Board],
    "Portfolio": [e.value for e in Portfolio],
    "SignalAction": [e.value for e in SignalAction],
    "TradeSide": [e.value for e in TradeSide],
    "TradeReason": [e.value for e in TradeReason],
    "LogLevel": [e.value for e in LogLevel],
    "OnboardingStep": [e.value for e in OnboardingStep],
    "ModelType": [e.value for e in ModelType],
    "ModelStatus": [e.value for e in ModelStatus],
}
